### PR TITLE
add zettelkasten module

### PR DESCRIPTION
**Pull request by ner0-m**
Don't get to exited this is not even alpha yet ;-)

I've mostly been trying to get familiar with neorg and the codebase, and put some things in place which will be needed in some for or the other. So basic creation of a new zettel is implemented and opening a zettel with a known id or a random is. Nothing more.

There are many areas which would need work. This is just a very rough summary of what has been discussed on discord the last couple of minutes:

- Goals
  - Links
    Links should use topic (first heading1?) as link text
    - Follow
       - follow links (already implemented in neorg)
       - splits
       - directly
    - Completion
      - get completion for links to other zettels
       - by name (topic?)
       - zettel number
    - Preview
      - allow to preview other zettels
      - just preview document.meta `description|
      - split
      - float
      - virtual text?
    - Backlinks
      There must a way of *automatic| backlinking:
     - In zettel **a** you create a link to zettel **b** > in zettel **b**< there will be a subsection `backlinks| with a link to $a|
  - Tags
   Use the categories from the document.meta
    - Searching, allow searching by categories
    - Quick adding, provide a quick way to add categories to a Zettel

- Structure
  - Zettel
    - Document meta
    -- Categories: used to search
    -- description: can be displayed for links
    - Heading 1: is used as title for zettel (link text)
    - Actual text
    - Backlinks
  - Zettelkasten
    - all the zettels, can be in folders
    - and index file with all zettels listed
    - file where zettels are sorted with categories
    - what do we do with zettels with multiple categories?

- Implementation Ideas
  - Linking
    - Database
      - sqlite.nvim
      - real database like org-roam v2

### CODE REVIEW

File: lua/neorg/modules/core/zettelkasten/module.lua [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+module.public = {
+    version = "0.0.1",
+
+    create_zettel = function(title)
+        -- Generate id/prefix
+        local prefix = module.config.private.id_fn()
+
+        -- Remove leading and trailing whitespace
+        local stripped_title = string.lower(string.gsub(title, "^%s*(.-)%s*$", "%1"))
+        local filename = string.format("%s-%s", prefix, string.gsub(stripped_title, "%s+", "-"))
```
Comment (danymat) [inline lua/neorg/modules/core/zettelkasten/module.lua:197]: I think the user should be able to specify if he wants `-` or whitespaces

File: lua/neorg/modules/core/zettelkasten/module.lua [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+        module.private.inject_template({ title = title })
+    end,
+
+    open_zettel_by_id = function(id)
+        local zettels = module.required["core.norg.dirman"].get_norg_files(module.config.public.workspace)
+
+        for _, zettel in ipairs(zettels) do
+            -- TODO: Can they have a path prefix, which I should remove?
+            local zid, _ = string.match(zettel, "([^-]+)-(.*).norg")
```
Comment (danymat) [inline lua/neorg/modules/core/zettelkasten/module.lua:210]: If allowing whitespaces, change this

**Comment by danymat** [comment]
One thought I have about the edit command, is we can override it with a telescope picker, in order to find all ZK files in the workspace

**Comment by danymat** [comment]
@ner0-m We deleted unstable branch, in favor of having only `main` branch. Can you reopen the issue by pointing the PR to main `instead` ?

**Comment by vhyrro** [comment]
Oh this autoclosed for some reason. I know that we had several issues that were supposed to autoclose due to the merge but not PRs lol. Wonder what happened there

**Comment by aaronclimbs** [comment]
Is this still being worked on? I think this would be a fantastic addition to the project.

**Comment by danymat** [comment]
Hello , it's being worked on by @max397574  on his repo, but we want Ro ship neorg 0.1 before commuting to this feature fully !
You can take a look at his code, and feel free to contribute !

**Pull request closed without merging**